### DISCUSSION TITLE
How to remove button after a certain time from view along with changing embed?

### DISCUSSION THREAD

**Discussion by ghost** [nextcord/nextcord#1102] [General]
This is my Buttons.py
```python
import nextcord

class EmbedButtons(nextcord.ui.View):
    @nextcord.ui.button(label="🎉 Enter giveaway", custom_id="enter_button")
    async def enter_callback(self, button, interaction: nextcord.Interaction):
        await interaction.response.send_message("Button was pressed", ephemeral=True)
```
This is another file called Bot.py and I am importing this, and running it in giveaway function of nextcord.Client.
```python
...
embed = nextcord.Embed(title="Giveaway is here", description=f"Giveaway")
await channel.send(embed=embed, view=EmbedButtons(timeout=None))
...
```
When someone clicks button, I log their userid via on_interaction() function checking bunch of message ids etc and finally adding it to a class variable (joined_users which is a list) of nextcord.Client. Now, I after x time lets say 600 seconds, I want to edit both embed to have list of joined users and view to remove the button (not disable it). I am not talking about on_timeout or any methods to acheive it in nextcord.ui.View class. I want to get callback with `interaction` as variable on nextcord.Client such that I can edit both embeds and views. How I can get such callback similar to on_interaction() of nextcord.Client

If thats not possible, is there any hacky way of doing what I want like passing embed and views to nextcord.ui.View, running on_timeout function then returning it to nextcord.Client then re-editing the message to include without button

**Comment by DenverCoder1** [comment]
You can remove the button from the view with [view.remove_item()](https://docs.nextcord.dev/en/stable/api.html#nextcord.ui.View.remove_item) or remove it from `view.children` and then edit the message passing the new view to the `view` parameter of message.edit or interaction.edit_message, or whatever method you end up using for editing.

If you want to remove the entire view, you can simply passing `view=None` should work.

**Reply by ghost** [reply]
Yes, but where to use view.remove_item()? Does, on_message or on_interaction receive an message or interaction when button timeouts if i set timeout=600? That is what I am facing problem at. @DenverCoder1

**Reply by DenverCoder1** [reply]
Sounds like you're looking for View.on_timeout?

```py
class ExampleView(nextcord.ui.View):
    def __init__(self):
        super().__init__(timeout=600)

    async def on_timeout(self):
        self.clear_items()
        await self.message.edit(view=self)
```

You can override that method how you like within the view and it will run when the view times out

**Reply by ghost** [reply]
@DenverCoder1 Thanks, just help me with last question please. So, lets say I managed to do embeds and view stuff. I managed to get list of users joined giveaway stored as nextcord.Client.joined_users. I also have other function called nextcord.Client.award_giveaway(userid). How do I get some function to triggered in nextcord.Client since self of on_timeout is self of nextcord.ui.View which doesn't have award_giveaway and joined_users variable? I want to award once button timeouts, it has to be automated, I saw some other bot implementing similar functionality so I think its possible

**Reply by DenverCoder1** [reply]
You can add parameters to the init function of the view to support passing the client variable and store it in self.client for example. Then, in any function of the view, you'll have access to the client with self.client.

```py
class ExampleView(nextcord.ui.View):
    def __init__(self, client):
        self.client = client
        super().__init__(timeout=600)

    async def on_timeout(self):
        # use self.client....

# instantiate with ExampleView(client)
```

Most of the time you shouldn't need client variables, storing the joined users and methods as fields in the view class should be good enough for most purposes.

**Reply by ghost** [reply]
ok thats an hacky way of doing it but I got it. EmbedsButton(timeout=600, myvar=1, myvar=2) right?. Ok I saw you edit it. Thanks for help

**Reply by ghost** [reply]
@DenverCoder1 
```
Traceback (most recent call last):
    await self.message.edit(view=self)
AttributeError: 'EmbedButtons' object has no attribute 'message'
```
This is the code. I haven't passed it Client variable yet, was just testing your upper code. Should I pass it message variable then run the command?
```

    async def on_timeout(self):
        self.clear_items()
        await self.message.edit(view=self)
```

**Reply by ghost** [reply]
Passed it message variable and it worked.

**Comment by ghost** [comment]
Resolved
